torch/csrc/distributed/rpc/init.cpp

```diff
@@ -290,6 +292,54 @@ If the future completes with an error, an exception is thrown.
       py::arg("tensors"),
       py::arg("rf") = nullptr);
 
+  // TODO This python future wrapper wraps c10::ivalue::Future.
+  // Will merge with JIT PythonFutureWrapper while merging generic Future with
+  // c10::ivalue::Future later on.
+  struct PythonFutureWrapper {
+    explicit PythonFutureWrapper(c10::intrusive_ptr<c10::ivalue::Future> fut)
+        : fut(std::move(fut)) {}
+
+    c10::intrusive_ptr<c10::ivalue::Future> fut;
+  };
+
+  // Since FutureMessage is binded to Future, here we need to bind the
+  // PythonFutureWrapper to a different name.
+  // TODO Once python object can be tagged as IValue and c10::ivalue::Future is
+  // implemented as generic Future<IValue>, we can consider all rpc call
+  // to return a future<IValue> later on.
+  shared_ptr_class_<PythonFutureWrapper>(module, "_pyFuture")
+      .def("wait", [](PythonFutureWrapper& fut) {
+        fut.fut->wait();
+        auto res = fut.fut->value();
+        {
+          // acquiring GIL as torch::jit::toPyObject creates new py::object
+          // without grabbing the GIL.
+          AutoGIL ag;
+          return torch::jit::toPyObject(std::move(res));
+        }
+      });
+
+  module.def(
+      "_invoke_rpc_script",
+      [](const std::string& dst,
+         const std::string& qualifiedName,
+         const py::args& args,
+         const py::kwargs& kwargs) {
+        // No need to catch exception here, if function can not be found,
+        // exception will be thrown in get_function() call; if args do not match
+        // with function schema, exception will be thrown in
+        // createStackForSchema() call.
+        auto name = c10::QualifiedName(qualifiedName);
+        auto fnSchema = PythonRpcHandler::getInstance()
+                            .jitCompilationUnit()
+                            ->get_function(name)
+                            .getSchema();
+        auto stack = torch::jit::createStackForSchema(
```

Comment:
After all, why do we need to put JIT-related work code in Python binding code?
I believe the python binding code should be thin, and not to bring in implementation details.

Reply:
we need to make rpcTorchscriptCall() return ivalue::Future, so that it can be called in JIT later on. here other codes are not JIT-related, it is actually python related codes that convert python args to IValues and wrap ivalue future to pyFuture